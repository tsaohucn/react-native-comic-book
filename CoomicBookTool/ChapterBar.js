import React, { Component }  from 'react'
import { 
  View, 
  StyleSheet, 
  Text, 
  Animated, 
  Dimensions, 
  TouchableWithoutFeedback ,
  FlatList
} from 'react-native'

export default class ProgressBar extends Component {

  constructor(props) {
    super(props)
    this.state = {
      currentChapterNumber: null
    }
  }

  receivePageNumber = pageNumber => {
    const chapterPageNumbers = this.props.chapter.map(ele => ele.pageNumber).filter(chapterPageNumber => chapterPageNumber <= pageNumber)
    const currentChapterNumber = chapterPageNumbers[chapterPageNumbers.length - 1]
    if (this.state.currentChapterNumber != currentChapterNumber) {
      this.setState({
        currentChapterNumber: currentChapterNumber
      })
    }
  }

  render() {
    return(
      <Animated.View style={[styles.chapterBarAnimated,{
        transform: [
          { translateX: this.props.animatedChapterBarX }
        ]}]}
      >
        <FlatList
          data={this.props.chapter.sort((a, b) => a.pageNumber - b.pageNumber)}
          keyExtractor={item => item.pageNumber.toString()}
          extraData={this.state.currentChapterNumber}
          showsVerticalScrollIndicator={false}
          renderItem={({ item }) => 
            <TouchableWithoutFeedback
              onPress={() => {
                this.props.onClickChapterItem && this.props.onClickChapterItem(item.pageNumber)
              }}
            >
              <View style={[styles.chapterBarView,{ backgroundColor: item.pageNumber === this.state.currentChapterNumber ? 'rgba(78, 78, 78, 0.8)' : 'rgba(52, 52, 52, 0.8)' }]}>
                <Text style={styles.chapterBarText}>{item.title}</Text>
                <Text style={styles.chapterBarText}>{item.pageNumber}</Text>
              </View>
            </TouchableWithoutFeedback>
          }
        />
      </Animated.View>
    )
  }
}

const { width, height } = Dimensions.get('window')

const chapterBarWidth = width*2/3

const styles = StyleSheet.create({
  chapterBarAnimated: {
    position: 'absolute', 
    top: 0,
    right: 0,
    height,
    width: chapterBarWidth,
    backgroundColor: 'rgba(52, 52, 52, 0.8)'
  },
  chapterBarView: {
    height: height/9,
    borderBottomColor: 'rgba(192,192,192,0.3)',
    borderBottomWidth: 0.3,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  chapterBarText: {
    fontSize: 15,
    color: 'white',
    padding: 10
  }
})