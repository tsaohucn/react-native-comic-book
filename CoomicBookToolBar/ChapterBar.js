import React from 'react'
import { 
  View, 
  StyleSheet, 
  Text, 
  Animated, 
  Dimensions, 
  TouchableWithoutFeedback ,
  FlatList
} from 'react-native'

const ChapterBar = ({ animatedChapterBarX, onClickChapterItem, chapter }) => (
  <Animated.View style={[styles.chapterBarAnimated,{
    transform: [
      { translateX: animatedChapterBarX }
    ]}]}
  >
    <FlatList
      data={chapter}
      renderItem={({ item }) => 
        <TouchableWithoutFeedback
          onPress={() => {
            onClickChapterItem && onClickChapterItem(item.pageNumber)
          }}
        >
          <View style={styles.chapterBarView}>
            <Text style={styles.chapterBarText}>{item.title}</Text>
          </View>
        </TouchableWithoutFeedback>
      }
    />
  </Animated.View>
)

const { width, height } = Dimensions.get('window')

const chapterBarWidth = width*2/3

const styles = StyleSheet.create({
  chapterBarAnimated: {
    position: 'absolute', 
    top: 0,
    right: 0,
    height,
    width: chapterBarWidth,
    backgroundColor: 'rgba(52, 52, 52, 0.8)',
    paddingLeft: 10
  },
  chapterBarView: {
    height: height/9,
    borderBottomColor: 'rgba(192,192,192,0.3)',
    borderBottomWidth: 0.3,
    justifyContent: 'center'
  },
  chapterBarText: {
    fontSize: 15,
    color: 'white'
  }
})

export default ChapterBar