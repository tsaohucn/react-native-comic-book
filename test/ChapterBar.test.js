// __tests__/ChapterBar.js
import 'react-native'
import React from 'react'
import ChapterBar from '../CoomicBookTool/ChapterBar'

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <ChapterBar />
    ).toJSON();
  expect(tree).toMatchSnapshot();
});