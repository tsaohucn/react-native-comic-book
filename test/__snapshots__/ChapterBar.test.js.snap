// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={undefined}
  style={
    Object {
      "backgroundColor": "rgba(52, 52, 52, 0.8)",
      "height": 1334,
      "paddingLeft": 10,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateX": undefined,
        },
      ],
      "width": 500,
    }
  }
>
  <RCTScrollView
    data={undefined}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    invertStickyHeaders={undefined}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
